openssl/rsa.go

```diff
@@ -64,26 +52,36 @@ func NewPublicKeyRSA(N, E *big.Int) (*PublicKeyRSA, error) {
 	n = bigToBN(N)
 	e = bigToBN(E)
 	C.go_openssl_RSA_set0_key(key, n, e, nil)
-	k := &PublicKeyRSA{_key: key}
+	pkey := C.go_openssl_EVP_PKEY_new()
+	if pkey == nil {
+		C.go_openssl_RSA_free(key)
+		return nil, newOpenSSLError("EVP_PKEY_new failed")
+	}
+	if C.go_openssl_EVP_PKEY_assign(pkey, C.EVP_PKEY_RSA, (unsafe.Pointer)(key)) != 1 {
+		C.go_openssl_RSA_free(key)
+		C.go_openssl_EVP_PKEY_free(pkey)
+		return nil, newOpenSSLError("EVP_PKEY_assign failed")
+	}
+	k := &PublicKeyRSA{_pkey: pkey}
 	runtime.SetFinalizer(k, (*PublicKeyRSA).finalize)
 	return k, nil
```

Comment:
When does `key` get freed on this code path? It doesn't seem to be returned and the other paths out of this function free the key. The doc for `EVP_PKEY_free` doesn't indicate that it will free this value.

Reply:
The [EVP_PKEY_assign*](https://beta.openssl.org/docs/manmaster/man3/EVP_PKEY_assign_RSA.html) docs indicate that `key` will be freed once `pkey` is freed.

> EVP_PKEY_assign_RSA(), EVP_PKEY_assign_DSA(), EVP_PKEY_assign_DH(), EVP_PKEY_assign_EC_KEY(), EVP_PKEY_assign_POLY1305() and EVP_PKEY_assign_SIPHASH() set the referenced key to key however these use the supplied key internally and so key will be freed when the parent pkey is freed.

They also mention that `EVP_PKEY_assign` is deprecated in OpenSSL 3, but changing it would suppose a huge divergence  between our OpenSSL 1 and 3 implementation, which is out of the scope for this PR (I have it implemented in https://github.com/microsoft/go-crypto-openssl/tree/dev/qmuntal/openssl3). I still have to understand the benefits of using the new `EVP_PKEY_fromdata` instead of the deprecated `EVP_PKEY_assign`. Both make use of the EVP interface, but the former works better with the new providers design.